### PR TITLE
Allow users to estimate query cost before executing it

### PR DESCRIPTION
**Pull request by betodealmeida**
### CATEGORY

Choose one

- [ ] Bug Fix
- [X] Enhancement (new features, refinement)
- [ ] Refactor
- [ ] Add tests
- [ ] Build / Development Environment
- [ ] Documentation

### SUMMARY

We currently added to Presto support for estimating the number of bytes scanned (https://github.com/prestosql/presto/pull/806), and we'd like to surface that information to SQL Lab users before they actually run a query.

This PR extends the DB specs with an `allows_cost_estimate` attribute and associated methods, allowing pre-execution costs to be computed from DBs that support it.

In order to use it, the feature flag `ESTIMATE_QUERY_COST` must be enabled, and it needs to be explicitly turned on for each database that supports query cost estimation. When all those conditions are met, a new button will show up in SQL Lab, allowing users to run cost estimates for the whole query or for the selected SQL.

### BEFORE/AFTER SCREENSHOTS OR ANIMATED GIF

DBs where the feature is not supported or not enabled are unmodified:

Screen Shot 2019-09-04 at 8 48 46 AM

Here's a Presto DB:

Screen Shot 2019-09-04 at 8 59 15 AM

Waiting for results:

query_cost

The result:

Screen Shot 2019-09-04 at 8 58 57 AM

And how errors (timeout, syntax errors) are surfaced:

Screen Shot 2019-09-04 at 8 53 07 AM

### TEST PLAN

Tested with a Presto cluster that supports query cost estimation, running version 0.319 and with the feature enabled via extra:

```json
{
    "version": "0.319",
    "cost_estimate_enabled": true,
    "metadata_params": {},
    "engine_params": {
       "connect_args": {
           "protocol":"https", 
           "source":"superset"
        }
    },
    "metadata_cache_timeout": {"schema_cache_timeout": 86400, "table_cache_timeout": 86400},
    "default_schemas": ["core", "default"]
}
```

Additionally, I tested:

- Lowering the Presto version removes the button.
- Other DBs don't show the button.
- Turning off the feature flag removes the button.

### ADDITIONAL INFORMATION

- [ ] Has associated issue:
- [X] Changes UI
- [ ] Requires DB Migration.
- [ ] Confirm DB Migration upgrade and downgrade tested.
- [X] Introduces new feature or API
- [ ] Removes existing feature or API

### REVIEWERS

@etr2460 @mistercrunch

### CODE REVIEW

File: superset/assets/src/SqlLab/components/EstimateQueryCostButton.jsx [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+      <span className="EstimateQueryCostButton">
+        <ModalTrigger
+          ref={this.queryCostModal}
+          modalTitle={t('Query Cost Estimate')}
+          modalBody={this.renderModalBody()}
+          triggerNode={
+            <Button
+              bsStyle="warning"
+              bsSize="small"
+              onClick={this.onClick}
```
Comment (etr2460) [inline superset/assets/src/SqlLab/components/EstimateQueryCostButton.jsx]: we probably need some type of indication that this is loading when you click on it, since i'll bet presto doesn't always return super fast.

File: superset/db_engine_specs/base.py [hunk 2]
```diff
@@ -643,6 +645,34 @@ def select_star(
             sql = sqlparse.format(sql, reindent=True)
         return sql
 
+    @classmethod
+    def estimate_statement_cost(cls, statement, database, cursor, user_name):
```
Comment (etr2460) [inline superset/db_engine_specs/base.py]: doc blocks and types please!

File: superset/db_engine_specs/base.py [hunk 3]
```diff
@@ -643,6 +645,34 @@ def select_star(
...
+        raise NotImplementedError("Subclasses should implement estimate_query_cost")
+
+    @classmethod
+    def estimate_query_cost(cls, database, schema, sql, source=None):
```
Comment (etr2460) [inline superset/db_engine_specs/base.py]: doc blocks and types please!

File: superset/db_engine_specs/presto.py [hunk 4]
```diff
@@ -373,6 +377,55 @@ def select_star(
             presto_cols,
         )
 
+    @classmethod
+    def estimate_statement_cost(cls, statement, database, cursor, user_name):
```
Comment (etr2460) [inline superset/db_engine_specs/presto.py]: doc blocks and types please!

File: superset/db_engine_specs/presto.py [hunk 5]
```diff
@@ -373,6 +377,55 @@ def select_star(
...
+
+        def humanize(value, suffix):
+            if value == "NaN":
+                return value
+
+            prefix = ""
+            symbols = ["K", "M", "G", "T", "P", "E", "Z", "Y"]
+            while value > 1000 and symbols:
+                prefix = symbols.pop(0)
+                value //= 1000
```
Comment (etr2460) [inline superset/db_engine_specs/presto.py]: shouldn't this be 1024? Also maybe we should make this magic number a constant

File: superset/db_engine_specs/presto.py [hunk 6]
```diff
@@ -373,6 +377,55 @@ def select_star(
...
+            sql = SQL_QUERY_MUTATOR(sql, user_name, security_manager, database)
+
+        sql = f"EXPLAIN (TYPE IO, FORMAT JSON) {sql}"
+
+        db_engine_spec.execute(cursor, sql)
+        polled = cursor.poll()
+        while polled:
+            time.sleep(0.2)
+            polled = cursor.poll()
+        data = db_engine_spec.fetch_data(cursor, 1)
```
Comment (etr2460) [inline superset/db_engine_specs/presto.py]: can you add a comment showing what this response looks like? It's not super obvious from the code how the result is being pulled out

File: superset/db_engine_specs/presto.py [hunk 7]
```diff
@@ -25,23 +25,27 @@
...
+from superset.sql_parse import ParsedQuery
 from superset.utils import core as utils
 
 QueryStatus = utils.QueryStatus
+config = app.config
 
 
 class PrestoEngineSpec(BaseEngineSpec):
     engine = "presto"
+    allows_cost_estimate = True
```
Comment (etr2460) [inline superset/db_engine_specs/presto.py]: What version of presto was this added in? Do we need to check the version number to make sure this is available?

File: superset/db_engine_specs/presto.py [hunk 8]
```diff
@@ -373,6 +377,55 @@ def select_star(
...
+            prefix = ""
+            symbols = ["K", "M", "G", "T", "P", "E", "Z", "Y"]
+            while value > 1000 and symbols:
+                prefix = symbols.pop(0)
+                value //= 1000
+
+            return f"{value} {prefix}{suffix}"
+
+        cost = {}
+        columns = [
```
Comment (etr2460) [inline superset/db_engine_specs/presto.py]: What happens if the presto format changes? Will we need to make changes here? I know presto likes to add a bunch of breaking changes in new versions...

File: superset/migrations/versions/8786d6374caa_add_column_for_query_estimate.py [hunk 9]
```diff
@@ -0,0 +1,40 @@
+# Licensed to the Apache Software Foundation (ASF) under one
```
Comment (etr2460) [inline superset/migrations/versions/8786d6374caa_add_column_for_query_estimate.py]: Is there a reason we wouldn't want to enable this for all supported dbs if the feature flag is turned on? It seems like the db migration might be overkill here. More of a design question tho.

File: superset/models/core.py [hunk 10]
```diff
@@ -773,6 +775,10 @@ def name(self):
     def allows_subquery(self):
         return self.db_engine_spec.allows_subqueries
 
+    @property
+    def allows_cost_estimate(self):
+        return self.db_engine_spec.allows_cost_estimate
```
Comment (etr2460) [inline superset/models/core.py]: with the current implementation, shouldn't this be:
`return self.cost_estimate_enabled and self.db_engine_spec.allows_cost_estimate`

File: superset/sql_validators/presto_db.py [hunk 11]
```diff
@@ -157,7 +157,7 @@ def validate(
         annotations: List[SQLValidationAnnotation] = []
         with closing(engine.raw_connection()) as conn:
             with closing(conn.cursor()) as cursor:
-                for statement in parsed_query.get_statements():
+                for statement in statements:
```
Comment (etr2460) [inline superset/sql_validators/presto_db.py]: this could be in a separate PR right? Might be better to pull it out since it's unrelated to this set of work

File: superset/views/core.py [hunk 12]
```diff
@@ -105,8 +106,12 @@
     get_viz,
 )
 
+
 config = app.config
 CACHE_DEFAULT_TIMEOUT = config.get("CACHE_DEFAULT_TIMEOUT", 0)
+SQLLAB_QUERY_COST_ESTIMATE_TIMEOUT = config.get(
```
Comment (etr2460) [inline superset/views/core.py]: not a big fan of configuring different timeouts for each different API call, is there a reason this is needed?

File: superset/views/core.py [hunk 13]
```diff
@@ -2391,6 +2396,36 @@ def select_star(self, database_id, table_name, schema=None):
...
+    @expose("/estimate_query_cost/<database_id>/", methods=["POST"])
+    @expose("/estimate_query_cost/<database_id>/<schema>/", methods=["POST"])
+    @event_logger.log_this
+    def estimate_query_cost(self, database_id, schema=None):
+        sql = json.loads(request.form.get("sql"))
+        template_params = json.loads(request.form.get("templateParams") or "{}")
+
+        if len(template_params) > 0:
+            return json_error_response(
+                "Query cost estimation does not support template parameters", status=400
```
Comment (etr2460) [inline superset/views/core.py]: why doesn't it support template parameters? Couldn't we just resolve the templates then run the estimation?

**Review by etr2460 — CHANGES_REQUESTED** [review]
a bunch of comments, also @john-bodley or @villebro should probably double check the db engine specs stuff

File: superset/assets/src/SqlLab/components/EstimateQueryCostButton.jsx (hunk 1 above)
Comment (betodealmeida) [reply superset/assets/src/SqlLab/components/EstimateQueryCostButton.jsx]: It does, it shows a loading wheel.

File: superset/db_engine_specs/base.py (hunk 2 above)
Comment (betodealmeida) [reply superset/db_engine_specs/base.py]: +1

File: superset/db_engine_specs/base.py (hunk 3 above)
Comment (betodealmeida) [reply superset/db_engine_specs/base.py]: +1

File: superset/db_engine_specs/presto.py (hunk 4 above)
Comment (betodealmeida) [reply superset/db_engine_specs/presto.py]: +1

File: superset/db_engine_specs/presto.py (hunk 5 above)
Comment (betodealmeida) [reply superset/db_engine_specs/presto.py]: You're right, I'll fix it.

File: superset/db_engine_specs/presto.py (hunk 6 above)
Comment (betodealmeida) [reply superset/db_engine_specs/presto.py]: Will do!

File: superset/db_engine_specs/presto.py (hunk 7 above)
Comment (betodealmeida) [reply superset/db_engine_specs/presto.py]: The PR has been merged, but It hasn't been released yet. We patched our Presto cluster with the commit and have it live.

Also, I'm not sure it's trivial to have features depending on the DB version, I remember asking @john-bodley about this for Druid.

File: superset/db_engine_specs/presto.py (hunk 8 above)
Comment (betodealmeida) [reply superset/db_engine_specs/presto.py]: Changes won't break this, but might render the response empty. Not sure how to best deal with this.

File: superset/models/core.py (hunk 10 above)
Comment (betodealmeida) [reply superset/models/core.py]: I preferred to implement that logic in the frontend, exposing the 2 separate flags.

File: superset/migrations/versions/8786d6374caa_add_column_for_query_estimate.py (hunk 9 above)
Comment (betodealmeida) [reply superset/migrations/versions/8786d6374caa_add_column_for_query_estimate.py]: I think it might be possible to have users running different versions of Presto (prod and staging, eg), so enabling the feature per DB seems to make sense.

File: superset/sql_validators/presto_db.py (hunk 11 above)
Comment (betodealmeida) [reply superset/sql_validators/presto_db.py]: Yeah, I was looking at this method and saw that `statements` was already computed. I'll write a separate PR.

File: superset/views/core.py (hunk 12 above)
Comment (betodealmeida) [reply superset/views/core.py]: My impression is that here we want a shorter timeout than usual. The query should return quickly, and if it doesn't return in a few seconds there's probably something wrong and it's not worth waiting for.

File: superset/views/core.py (hunk 13 above)
Comment (betodealmeida) [reply superset/views/core.py]: The template processing is tightly coupled with the database (for resolving macros related to latest partition, etc.), and it was hard to bring it on here.

File: superset/assets/src/SqlLab/components/EstimateQueryCostButton.jsx (hunk 1 above)
Comment (etr2460) [reply superset/assets/src/SqlLab/components/EstimateQueryCostButton.jsx]: could you attach a gif to show the behavior?

File: superset/db_engine_specs/presto.py (hunk 7 above)
Comment (etr2460) [reply superset/db_engine_specs/presto.py]: I implemented this already in another part of the presto engine spec, relying on the user specifying the version in the extra data field of the db model, but falling back to assuming the most recent version. check out: https://github.com/apache/incubator-superset/blob/master/superset/db_engine_specs/presto.py#L918

File: superset/migrations/versions/8786d6374caa_add_column_for_query_estimate.py (hunk 9 above)
Comment (etr2460) [reply superset/migrations/versions/8786d6374caa_add_column_for_query_estimate.py]: As mentioned above, maybe enable per DB through using the extra data version instead?

File: superset/db_engine_specs/presto.py (hunk 8 above)
Comment (etr2460) [reply superset/db_engine_specs/presto.py]: I guess we'll just need to keep an eye on it

File: superset/assets/src/SqlLab/components/EstimateQueryCostButton.jsx (hunk 1 above)
Comment (betodealmeida) [reply superset/assets/src/SqlLab/components/EstimateQueryCostButton.jsx]: Done.

File: superset/db_engine_specs/presto.py (hunk 5 above)
Comment (betodealmeida) [reply superset/db_engine_specs/presto.py]: Actually, it should be 1000: https://en.wikipedia.org/wiki/Kibibyte

I'll put it in a constant.

Comment (betodealmeida) [reply superset/db_engine_specs/presto.py]: Also, the CPU cost is not in bytes (it's unitless), so we should use 1000.

File: superset/db_engine_specs/presto.py (hunk 8 above)
Comment (betodealmeida) [reply superset/db_engine_specs/presto.py]: This is a feature we're expecting to use a lot, so we'll have a good number of eyes. But ideally we'd have integration tests with Presto.

File: superset/db_engine_specs/presto.py (hunk 7 above)
Comment (betodealmeida) [reply superset/db_engine_specs/presto.py]: Ah, nice. Let me take a look.

File: superset/views/core.py (hunk 13 above)
Comment (betodealmeida) [reply superset/views/core.py]: Actually I got this to work.

File: superset/db_engine_specs/presto.py (hunk 7 above)
Comment (betodealmeida) [reply superset/db_engine_specs/presto.py]: Cool, I'll remove the DB migration and make it configurable via extras, as well as using the version to see if the DB supports cost estimation.

File: superset/migrations/versions/8786d6374caa_add_column_for_query_estimate.py (hunk 9 above)
Comment (betodealmeida) [reply superset/migrations/versions/8786d6374caa_add_column_for_query_estimate.py]: Will do.

**Comment by betodealmeida** [comment]
@etr2460, I addressed all the comments. I removed the DB migration, and the feature is enabled per DB in extras. Also, I use the DB version to determine if it's supported.

File: superset/views/core.py (hunk 12 above)
Comment (etr2460) [inline superset/views/core.py]: Can you add this config var to the config file and document what it does there?

File: superset/views/core.py [hunk 14]
```diff
@@ -2391,6 +2396,34 @@ def select_star(self, database_id, table_name, schema=None):
             mydb.select_star(table_name, schema, latest_partition=True, show_cols=True)
         )
 
+    @has_access_api
+    @expose("/estimate_query_cost/<database_id>/", methods=["POST"])
+    @expose("/estimate_query_cost/<database_id>/<schema>/", methods=["POST"])
+    @event_logger.log_this
+    def estimate_query_cost(self, database_id: int, schema: str = None) -> Response:
+        mydb = db.session.query(models.Database).filter_by(id=database_id).first()
```
Comment (etr2460) [inline superset/views/core.py]: i think that `one_or_none()` is a bit clearer than `first()`, but it's just personal preference

**Review by etr2460 — APPROVED** [review]
Would you mind making the test plan a little more robust? Test with the feature flag both enabled and disabled, with presto dbs that are configured at a passing version and prior version? With non presto dbs? I'm sure you've tested other cases, but right now the test plan only references the happy path, so a bit more detail would be great.

other than that and my 2 comments here, this lgtm! I'll approve to unblock

File: superset/views/core.py (hunk 12 above)
Comment (betodealmeida) [reply superset/views/core.py]: +1

File: superset/views/core.py (hunk 14 above)
Comment (betodealmeida) [reply superset/views/core.py]: Ah, I wasn't familiar with `one_or_none()`, seems better suited. I'll change.
